### PR TITLE
[SPARK-14346] [SQL][WIP]Show create table 

### PR DESCRIPTION
**Pull request by xwu0226**
## What changes were proposed in this pull request?
1. add `SHOW CREATE TABLE tableIdentifier`  in **SqlBase.g4** file
2. create a `RunnableCommand` class for `ShowCreateTableCommand`, that invokes `SessionCatalog.generateTableDDL`
3. add `generateTableDDL` method in `SessionCatalog`and override it in `HiveSessionCatalog`.
4. `HiveSessionCatalog.generateTableDDL` is to create DDL for metastore tables. and for tempTables in Spark, it may not be quite practical. For now, we throw exception in `SessionCatalog.generateTableDDL`. If the community thinks it is also necessary, we can implement this later. 
5. `HiveSessionCatalog.generateTableDDL` branches out to generate Hive syntax DDL for hive tables, while it is also supposed to branch out to generate DataSource syntax DDL for datasource tables. 
6. Limitation. While the DDL work has been going on, some datasource table features are not supported yet. So the generation of DDL is also limited. For example:
   If the table is created as`df.write.partitionBy("a").saveAsTable("t1")`, this datasource table will record the partitioning columns in tblproperties of the metastore table. But, when the datasource syntax DDL is generated based on this table, we will lose the partitioning information in the DDL. Similarly for bucketing and skew information. etc. 
7. Some follow-up PRs for TODOs:
- For some cases, where "create table" used to issue warnings and may throw exception instead, show create table need to adjust to it. 
- After datasource table DDL can support partitioning, bucketing, skewing, show create table DDL needs to support.
- Complete the support of SORTCOLUMNS for bucketing and StorageHandler inforamtion in CatalogTable class, then, we can support "SORTED BY", "SKEWED BY", "STORED BY" clause in showing Hive syntax DDL
- Support show create table index?
## How was this patch tested?

unite tests with self-check. run the hive, sql, catalyst sbt tests.

### CODE REVIEW

**Comment by gatorsmile** [comment]
Please change the title to ` [SPARK-14346] [SQL][WIP] Native Show Create Table Command` Thanks!

**Comment by andrewor14** [comment]
ok to test

**Comment by xwu0226** [comment]
@andrewor14 , Thank you for triggering off the test. I just pushed the fix for the scala style issues. Thanks again!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveShowDDLSuite.scala [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  test("show create table - hive table - no row format") {
+    withTempDir { tmpDir =>
+      withTable("t1") {
+        sql(
+          s"""
+            |create table t1(c1 int, c2 string)
+            |stored as parquet
+            |location '${tmpDir}'
+          """.stripMargin)
+        sql("show create table t1").show(false)
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveShowDDLSuite.scala]: Please do not forget to remove all these `show(false)`

**Comment by gatorsmile** [comment]
Please move these test cases from black_list to white_list in `HiveCompatibilitySuite`
    "show_create_table_alter",
    "show_create_table_db_table",
    "show_create_table_delimited",
    "show_create_table_does_not_exist",
    "show_create_table_index",
    "show_create_table_partitioned",
    "show_create_table_serde",
    "show_create_table_view",

Thanks!

**Comment by xwu0226** [comment]
@gatorsmile  Thanks, Xiao! I will do that.

**Comment by xwu0226** [comment]
After the CREATE TABLE command is supported natively by [SPARK-14388](https://github.com/apache/spark/pull/12271/), the logic in this PR will break and need to adjust, including testcases.

**Comment by xwu0226** [comment]
@yhuai After SPARK-14388 is just merged, I am starting to modify my logic for this PR to adapt.. One question is that, the result of native command `SHOW CREATE TABLE ...` has to be run successfully natively in SparkSQL, correct?  Then, we may lose some unsupported DDL features if the tables are originally created by Hive, such as bucketing, sorting, skew, etc.  Is this OK?

**Comment by yhuai** [comment]
If a generated CREATE TABLE statement cannot be executed in Spark, how about we mention it in the output?

**Comment by xwu0226** [comment]
@yhuai So the requirement is to have a generated DDL that has full DDL features for the target table, regardless of whether Spark can support it natively or not.  Then, I think we can add a comment row in the results of the `SHOW CREATE TABLE ...`command.

**Comment by yhuai** [comment]
Yea, we should avoid of generating a DDL that cannot creates the original table. If we put the message in the comment field, will it be noticed easily?

**Comment by xwu0226** [comment]
For the the message notifying that this DDL can not be run natively in Spark, I meant to output it as another column or row in the output of the `SHOW CREATE TABLE ... `such as:

DDL | Remarks
------ | --------
CREATE TABLE ....  CLUSTERED BY (.... SORTED BY (...))...| This DDL is not supported by Spark SQL natively, because ....

**Comment by yhuai** [comment]
oh, i see. Thanks. Let's start with it.

**Comment by xwu0226** [comment]
@yhuai Another question.. Currently, for the tables that has `spark.sql.sources.provider` in table properties, I generate DDL with datasource table syntax, such as `create table ... USING .. OPTIONS (...)`. However, such syntax does not support partitioning, bucketing, sorting features. We will lose such information, if the table is created as `df.write.partitionBy(...).bucketBy(...).saveAsTable(...)`
Should we also create hive DDL syntax for datasource table then? Thanks!

**Comment by yhuai** [comment]
oh i see. How about we also show how to create the table in DataFrame API for now?

**Comment by xwu0226** [comment]
So we have 3 cases:
1. Tables created with Hive syntax -- we generate DDL in hive syntax 
2. Tables created with datasource syntax -- we generate DDL in datasource syntax `...USING...OPTIONS(...)`
3. Tables created from dataframe -- If the table metadata contains the features that are not supported in datasource syntax, generate create table with dataframe API. The reason for this condition is that under the cover there is no indication of whether the table is created with dataframe API or not, it is just a datasource table by `spark.sql.sources.provider`. 
@yhuai what do you think? Thanks!

**Comment by xwu0226** [comment]
Somehow my merge brought in a lot of other stuff in this PR. I need to close this PR and resubmit another PR.

**Comment by gatorsmile** [comment]
A new PR has been created. https://github.com/apache/spark/pull/12406

All the comments are addressed there. Thanks!

**Pull request closed without merging**